File: go/pkg/voiceprint/model_ncnn.go
package voiceprint

import (
	"fmt"
	"math"
	"os"
	"sync"

	"github.com/haivivi/giztoy/go/pkg/ncnn"
)

const (
	// segFrames is the number of fbank frames per inference segment.
	// 300 frames = 3 seconds at 10ms hop. Matching the training window
	// of speaker embedding models improves stability.
	segFrames = 300

	// hopFrames is the hop between segments for averaging.
	hopFrames = 150
)

// NCNNModel implements [Model] using the ncnn inference engine for speaker
// embedding extraction. It wraps a ncnn.Net and handles the full pipeline
// from PCM audio to embedding vector.
//
// # Model Pipeline
//
//  1. PCM16 audio → [ComputeFbank] → mel filterbank features
//  2. Fbank features → ncnn inference → speaker embedding
//
// # Thread Safety
//
// NCNNModel is safe for concurrent use. The ncnn.Net is loaded once and
// shared; each Extract call creates its own ncnn.Extractor.
// Extract holds a read lock for the entire inference duration to prevent
// Close from destroying the net while inference is in progress.
type NCNNModel struct {
	mu       sync.RWMutex
	net      *ncnn.Net
	dim      int
	fbankCfg FbankConfig
	closed   bool

	inputName  string
	outputName string
}

// NCNNModelOption configures an NCNNModel.
type NCNNModelOption func(*NCNNModel)

// WithNCNNFbankConfig sets the filterbank configuration.
func WithNCNNFbankConfig(cfg FbankConfig) NCNNModelOption {
	return func(m *NCNNModel) {
		m.fbankCfg = cfg
	}
}

// WithNCNNEmbeddingDim overrides the expected embedding dimension.
// Default: 512 (for 3D-Speaker ERes2Net base model).
func WithNCNNEmbeddingDim(dim int) NCNNModelOption {
	return func(m *NCNNModel) {
		if dim > 0 {
			m.dim = dim
		}
	}
}

// WithNCNNBlobNames sets the input and output blob names.
// Default: "in0" and "out0" (PNNX-converted models).
func WithNCNNBlobNames(input, output string) NCNNModelOption {
	return func(m *NCNNModel) {
		m.inputName = input
		m.outputName = output
	}
}

// NewNCNNModel creates a new NCNNModel by loading .param and .bin from files.
// For embedding the model in the binary, use [NewNCNNModelFromMemory] instead.
//
// Note: FP16 is disabled by loading the model data from memory internally.
// This avoids the ncnn limitation where options must be set before loading.
func NewNCNNModel(paramPath, binPath string, opts ...NCNNModelOption) (*NCNNModel, error) {
	m := newNCNNModelDefaults(opts)

	// Read files and load via memory path to ensure FP16 is disabled before load.
	paramData, err := os.ReadFile(paramPath)
	if err != nil {
		return nil, fmt.Errorf("voiceprint: read param: %w", err)
	}
	binData, err := os.ReadFile(binPath)
	if err != nil {
		return nil, fmt.Errorf("voiceprint: read bin: %w", err)
	}

	opt := ncnn.NewOption()
	opt.SetFP16(false)
	net, err := ncnn.NewNetFromMemory(paramData, binData, opt)
	if err != nil {
		return nil, fmt.Errorf("voiceprint: %w", err)
	}
	m.net = net
	return m, nil
}

// NewNCNNModelFromMemory creates a new NCNNModel from in-memory .param and .bin data.
// This is the preferred constructor when the model is embedded in the binary
// via go:embed, producing a single-file deployment with zero external dependencies.
func NewNCNNModelFromMemory(paramData, binData []byte, opts ...NCNNModelOption) (*NCNNModel, error) {
	if len(paramData) == 0 {
		return nil, fmt.Errorf("voiceprint: empty param data")
	}
	if len(binData) == 0 {
		return nil, fmt.Errorf("voiceprint: empty bin data")
	}

	m := newNCNNModelDefaults(opts)

	opt := ncnn.NewOption()
	opt.SetFP16(false)
	net, err := ncnn.NewNetFromMemory(paramData, binData, opt)
	if err != nil {
		return nil, fmt.Errorf("voiceprint: %w", err)
	}
	m.net = net
	return m, nil
}

// NewNCNNModelFromNet creates a new NCNNModel from a pre-loaded ncnn.Net.
// Use this when loading models via [ncnn.LoadModel] from the model registry.
func NewNCNNModelFromNet(net *ncnn.Net, opts ...NCNNModelOption) *NCNNModel {
	m := newNCNNModelDefaults(opts)
	m.net = net
	return m
}

func newNCNNModelDefaults(opts []NCNNModelOption) *NCNNModel {
	m := &NCNNModel{
		dim:        512,
		fbankCfg:   DefaultFbankConfig(),
		inputName:  "in0",
		outputName: "out0",
	}
	for _, opt := range opts {
		opt(m)
	}
	return m
}

// Extract implements [Model]. It converts PCM16 audio to a normalized
// speaker embedding using the following pipeline:
//
//  1. PCM → fbank features (with Kaldi-compatible parameters)
//  2. CMVN normalization (removes channel/environment effects)
//  3. Segment-based inference (300-frame windows with 150-frame hop)
//  4. Average segment embeddings + L2 normalize
//
// The returned embedding is an L2-normalized unit vector suitable for
// cosine similarity comparison via dot product.
func (m *NCNNModel) Extract(audio []byte) ([]float32, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	if m.closed {
		return nil, fmt.Errorf("voiceprint: model is closed")
	}
	net := m.net

	// Step 1: Compute fbank features.
	features := ComputeFbank(audio, m.fbankCfg)
	if features == nil || len(features) == 0 {
		return nil, fmt.Errorf("voiceprint: audio too short for fbank extraction")
	}

	// Step 2: CMVN — subtract mean, divide by std per mel bin.
	cmvn(features)

	// Step 3: Segment-based extraction with averaging.
	numFrames := len(features)
	if numFrames <= segFrames {
		// Short audio: single inference.
		emb, err := m.extractSegment(net, features)
		if err != nil {
			return nil, err
		}
		l2Normalize(emb)
		return emb, nil
	}

	// Long audio: sliding window, average all segment embeddings.
	var embeddings [][]float32
	var lastStart int
	for start := 0; start+segFrames <= numFrames; start += hopFrames {
		emb, err := m.extractSegment(net, features[start:start+segFrames])
		if err != nil {
			continue
		}
		l2Normalize(emb)
		embeddings = append(embeddings, emb)
		lastStart = start
	}
	// Ensure the last segment covers the end of the audio.
	if tail := numFrames - segFrames; tail > lastStart {
		emb, err := m.extractSegment(net, features[tail:])
		if err == nil {
			l2Normalize(emb)
			embeddings = append(embeddings, emb)
		}
	}

	if len(embeddings) == 0 {
		return nil, fmt.Errorf("voiceprint: all segments failed")
	}

	// Average all segment embeddings.
	avg := make([]float32, m.dim)
	for _, emb := range embeddings {
		for i, v := range emb {
			avg[i] += v
		}
	}
	n := float32(len(embeddings))
	for i := range avg {
		avg[i] /= n
	}
	l2Normalize(avg)
	return avg, nil
}

// extractSegment runs ncnn inference on a single fbank segment.
func (m *NCNNModel) extractSegment(net *ncnn.Net, features [][]float32) ([]float32, error) {
	numFrames := len(features)
	numMels := len(features[0])

	flatData := make([]float32, numFrames*numMels)
	for t := 0; t < numFrames; t++ {
		copy(flatData[t*numMels:], features[t])
	}

	input, err := ncnn.NewMat2D(numMels, numFrames, flatData)
	if err != nil {
		return nil, fmt.Errorf("voiceprint: create input mat: %w", err)
	}
	defer input.Close()

	ex, err := net.NewExtractor()
	if err != nil {
		return nil, fmt.Errorf("voiceprint: create extractor: %w", err)
	}
	defer ex.Close()

	if err := ex.SetInput(m.inputName, input); err != nil {
		return nil, fmt.Errorf("voiceprint: %w", err)
	}

	output, err := ex.Extract(m.outputName)
	if err != nil {
		return nil, fmt.Errorf("voiceprint: %w", err)
	}
	defer output.Close()

	data := output.FloatData()
	if data == nil {
		return nil, fmt.Errorf("voiceprint: ncnn output data is nil")
	}

	n := len(data)
	if n > m.dim {
		n = m.dim
	}
	embedding := make([]float32, m.dim)
	copy(embedding, data[:n])
	return embedding, nil
}

// cmvn applies Cepstral Mean and Variance Normalization in-place.
// For each mel dimension, subtracts the mean and divides by the standard
// deviation across all frames. This removes channel and environment effects.
func cmvn(features [][]float32) {
	if len(features) == 0 {
		return
	}
	numMels := len(features[0])
	T := float64(len(features))

	for m := 0; m < numMels; m++ {
		var sum float64
		for _, f := range features {
			sum += float64(f[m])
		}
		mean := sum / T

		var varSum float64
		for _, f := range features {
			d := float64(f[m]) - mean
			varSum += d * d
		}
		std := math.Sqrt(varSum / T)
		if std < 1e-10 {
			std = 1e-10
		}

		for _, f := range features {
			f[m] = float32((float64(f[m]) - mean) / std)
		}
	}
}

// l2Normalize normalizes a vector to unit length in-place.
func l2Normalize(v []float32) {
	var norm float64
	for _, x := range v {
		norm += float64(x) * float64(x)
	}
	norm = math.Sqrt(norm)
	if norm > 0 {
		scale := float32(1.0 / norm)
		for i := range v {
			v[i] *= scale
		}
	}
}

// Dimension implements [Model].
func (m *NCNNModel) Dimension() int {
	return m.dim
}

// Close implements [Model].
func (m *NCNNModel) Close() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.closed {
		return nil
	}
	m.closed = true
	if m.net != nil {
		m.net.Close()
		m.net = nil
	}
	return nil
}

File: rust/voiceprint/src/model_ncnn.rs
//! [`VoiceprintModel`] implementation using the ncnn inference engine.

use std::sync::RwLock;

use giztoy_ncnn::{Mat, NcnnOption, Net};

use crate::error::VoiceprintError;
use crate::fbank::{cmvn, compute_fbank, l2_normalize, FbankConfig};
use crate::model::VoiceprintModel;

/// Number of fbank frames per inference segment.
/// 300 frames = 3 seconds at 10ms hop.
const SEG_FRAMES: usize = 300;

/// Hop between segments for averaging.
const HOP_FRAMES: usize = 150;

/// [`VoiceprintModel`] implementation using the ncnn inference engine.
///
/// Wraps a [`Net`] and handles the full pipeline from PCM audio to
/// speaker embedding vector.
///
/// # Pipeline
///
/// 1. PCM16 audio -> [`compute_fbank`] -> mel filterbank features
/// 2. Fbank features -> [`cmvn`] normalization
/// 3. Segment-based ncnn inference (300-frame windows, 150-frame hop)
/// 4. Average segment embeddings + L2 normalize
///
/// # Thread Safety
///
/// NCNNModel is safe for concurrent use. The ncnn Net is loaded once
/// and shared; each `extract` call creates its own Extractor.
pub struct NCNNModel {
    inner: RwLock<NCNNModelInner>,
}

struct NCNNModelInner {
    net: Option<Net>,
    dim: usize,
    fbank_cfg: FbankConfig,
    input_name: String,
    output_name: String,
    closed: bool,
}

/// Configuration for [`NCNNModel`].
pub struct NCNNModelConfig {
    /// Expected embedding dimension (default: 512).
    pub dim: usize,
    /// Filterbank configuration.
    pub fbank_cfg: FbankConfig,
    /// ncnn input blob name (default: "in0").
    pub input_name: String,
    /// ncnn output blob name (default: "out0").
    pub output_name: String,
}

impl Default for NCNNModelConfig {
    fn default() -> Self {
        Self {
            dim: 512,
            fbank_cfg: FbankConfig::default(),
            input_name: "in0".to_string(),
            output_name: "out0".to_string(),
        }
    }
}

impl NCNNModel {
    /// Creates a new NCNNModel from in-memory .param and .bin data.
    /// FP16 is disabled by default for numerical safety.
    pub fn from_memory(
        param_data: &[u8],
        bin_data: &[u8],
        cfg: NCNNModelConfig,
    ) -> Result<Self, VoiceprintError> {
        if param_data.is_empty() || bin_data.is_empty() {
            return Err(VoiceprintError::Model("empty model data".into()));
        }

        let mut opt = NcnnOption::new().map_err(|e| VoiceprintError::Model(e.to_string()))?;
        opt.set_fp16(false);
        let net = Net::from_memory(param_data, bin_data, Some(&opt))
            .map_err(|e| VoiceprintError::Model(e.to_string()))?;

        Ok(Self {
            inner: RwLock::new(NCNNModelInner {
                net: Some(net),
                dim: cfg.dim,
                fbank_cfg: cfg.fbank_cfg,
                input_name: cfg.input_name,
                output_name: cfg.output_name,
                closed: false,
            }),
        })
    }

    /// Creates a new NCNNModel from a pre-loaded Net.
    pub fn from_net(net: Net, cfg: NCNNModelConfig) -> Self {
        Self {
            inner: RwLock::new(NCNNModelInner {
                net: Some(net),
                dim: cfg.dim,
                fbank_cfg: cfg.fbank_cfg,
                input_name: cfg.input_name,
                output_name: cfg.output_name,
                closed: false,
            }),
        }
    }

    /// Closes the model and releases resources.
    pub fn close(&self) {
        let mut inner = self.inner.write().unwrap();
        if !inner.closed {
            inner.closed = true;
            inner.net = None;
        }
    }
}

impl VoiceprintModel for NCNNModel {
    fn extract(&self, audio: &[u8]) -> Result<Vec<f32>, VoiceprintError> {
        let inner = self.inner.read().unwrap();
        if inner.closed {
            return Err(VoiceprintError::Closed);
        }
        let net = inner.net.as_ref().ok_or(VoiceprintError::Closed)?;

        // Step 1: Compute fbank features.
        let mut features =
            compute_fbank(audio, &inner.fbank_cfg).ok_or(VoiceprintError::AudioTooShort {
                min_bytes: inner.fbank_cfg.frame_length * 2,
                got_bytes: audio.len(),
            })?;

        if features.is_empty() {
            return Err(VoiceprintError::AudioTooShort {
                min_bytes: inner.fbank_cfg.frame_length * 2,
                got_bytes: audio.len(),
            });
        }

        // Step 2: CMVN normalization.
        cmvn(&mut features);

        // Step 3: Segment-based extraction with averaging.
        let num_frames = features.len();
        if num_frames <= SEG_FRAMES {
            let mut emb = extract_segment(
                net,
                &features,
                &inner.input_name,
                &inner.output_name,
                inner.dim,
            )?;
            l2_normalize(&mut emb);
            return Ok(emb);
        }

        // Long audio: sliding window, average all segment embeddings.
        let mut embeddings: Vec<Vec<f32>> = Vec::new();
        let mut last_start = 0;
        let mut start = 0;
        while start + SEG_FRAMES <= num_frames {
            if let Ok(mut emb) = extract_segment(
                net,
                &features[start..start + SEG_FRAMES],
                &inner.input_name,
                &inner.output_name,
                inner.dim,
            ) {
                l2_normalize(&mut emb);
                embeddings.push(emb);
            }
            last_start = start;
            start += HOP_FRAMES;
        }

        // Ensure the last segment covers the end of the audio.
        let tail = num_frames - SEG_FRAMES;
        if tail > last_start {
            if let Ok(mut emb) = extract_segment(
                net,
                &features[tail..],
                &inner.input_name,
                &inner.output_name,
                inner.dim,
            ) {
                l2_normalize(&mut emb);
                embeddings.push(emb);
            }
        }

        if embeddings.is_empty() {
            return Err(VoiceprintError::Model("all segments failed".into()));
        }

        // Average all segment embeddings.
        let mut avg = vec![0.0f32; inner.dim];
        for emb in &embeddings {
            for (i, &v) in emb.iter().enumerate() {
                avg[i] += v;
            }
        }
        let n = embeddings.len() as f32;
        for v in &mut avg {
            *v /= n;
        }
        l2_normalize(&mut avg);
        Ok(avg)
    }

    fn dimension(&self) -> usize {
        self.inner.read().unwrap().dim
    }
}

/// Runs ncnn inference on a single fbank segment.
fn extract_segment(
    net: &Net,
    features: &[Vec<f32>],
    input_name: &str,
    output_name: &str,
    dim: usize,
) -> Result<Vec<f32>, VoiceprintError> {
    let num_frames = features.len();
    let num_mels = features[0].len();

    let mut flat_data = vec![0.0f32; num_frames * num_mels];
    for (t, frame) in features.iter().enumerate() {
        flat_data[t * num_mels..t * num_mels + num_mels].copy_from_slice(frame);
    }

    let input = Mat::new_2d(num_mels as i32, num_frames as i32, &flat_data)
        .map_err(|e| VoiceprintError::Model(format!("create input mat: {e}")))?;

    let mut ex = net
        .extractor()
        .map_err(|e| VoiceprintError::Model(format!("create extractor: {e}")))?;

    ex.set_input(input_name, &input)
        .map_err(|e| VoiceprintError::Model(e.to_string()))?;

    let output = ex
        .extract(output_name)
        .map_err(|e| VoiceprintError::Model(e.to_string()))?;

    let data = output
        .float_data()
        .ok_or_else(|| VoiceprintError::Model("ncnn output data is nil".into()))?;

    let n = data.len().min(dim);
    let mut embedding = vec![0.0f32; dim];
    embedding[..n].copy_from_slice(&data[..n]);
    Ok(embedding)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ncnn_model_dimension() {
        giztoy_ncnn::register_embedded_models();
        let net = giztoy_ncnn::load_model(giztoy_ncnn::ModelId::SPEAKER_ERES2NET).unwrap();
        let model = NCNNModel::from_net(net, NCNNModelConfig::default());
        assert_eq!(model.dimension(), 512);
    }

    #[test]
    fn ncnn_model_extract_short_audio() {
        giztoy_ncnn::register_embedded_models();
        let net = giztoy_ncnn::load_model(giztoy_ncnn::ModelId::SPEAKER_ERES2NET).unwrap();
        let model = NCNNModel::from_net(net, NCNNModelConfig::default());

        // 1 second of silence (16000 samples = 32000 bytes).
        let audio = vec![0u8; 32000];
        let emb = model.extract(&audio).unwrap();
        assert_eq!(emb.len(), 512);

        // Should be L2-normalized (unit length).
        let norm: f64 = emb
            .iter()
            .map(|&x| (x as f64) * (x as f64))
            .sum::<f64>()
            .sqrt();
        assert!(
            (norm - 1.0).abs() < 1e-4,
            "embedding should be unit length, got {norm}"
        );
    }

    #[test]
    fn ncnn_model_too_short_audio() {
        giztoy_ncnn::register_embedded_models();
        let net = giztoy_ncnn::load_model(giztoy_ncnn::ModelId::SPEAKER_ERES2NET).unwrap();
        let model = NCNNModel::from_net(net, NCNNModelConfig::default());

        // 100 bytes = 50 samples, too short for a frame.
        let audio = vec![0u8; 100];
        assert!(model.extract(&audio).is_err());
    }

    #[test]
    fn ncnn_model_close() {
        giztoy_ncnn::register_embedded_models();
        let net = giztoy_ncnn::load_model(giztoy_ncnn::ModelId::SPEAKER_ERES2NET).unwrap();
        let model = NCNNModel::from_net(net, NCNNModelConfig::default());
        model.close();
        assert!(model.extract(&vec![0u8; 32000]).is_err());
    }
}
